Use ObjectHandles for the pending message queue

livecodefraser opened the pull request:
This prevents crashes when objects with pending messages are deleted.

livecodefraser committed:
Make MCAutoValueRef<...>::Reset safe for self-assignment

livecodefraser committed:
Add a copy constructor to MCAutoValueRef<...>

livecodefraser committed:
Fix null ptr deref in MCObjptr::getrefasgroup

livecodefraser committed:
Refactor pending messages queue to use MCObjectHandles

Instead of raw object pointers. Previously, it was possible for the
engine to attempt to send a message to a deleted object, causing
sadness.

peter-b commented on engine/src/uidc.cpp:
```diff
@@ -1272,36 +1325,41 @@ Boolean MCUIDC::handlepending(real8& curtime, real8& eventtime, Boolean dispatch
-            MCU_restoreprops(sp);
-            while (p != NULL)
+            // Remove this message from the queue
+            cancelmessageindex(i, false);
```
I think you need to decrement `i` here in order to avoid skipping over the following message.

livecodefraser replied:
It does look rather suspicious, doesn't it?

livecodefraser committed:
Fix an error in handing pending messages

The counter was incremented when a message was removed, causing the
next message in the queue to be skipped. This appears to be a long-
standing bug!

livecodefraser committed:
Add constructors to MCPendingMessage

The C++11 support in MSVC 2010 is not advanced enough to use the brace-
initialisation syntax that I was using so specific constructors need to
be added instead. These can be removed again when we use a newer version
of the compiler.

livecodefraser committed:
Fix Win32-specific code to use new MCPendingMessageList code

xvi-fraser committed:
[[ Bugfix 18588 ]] Add bugfix note

peter-b commented:
@livecodefraser This is sort of crying out for a `MCPendingMessageIterator` 😉 

@livecode-vulcan review ok 9c3049c

livecode-vulcan commented:
💙 review by @peter-b ok 9c3049cfc9a9b6e525b3d4449b89edb91579d320

livecode-vulcan commented:
😎 test success 9c3049cfc9a9b6e525b3d4449b89edb91579d320
- try-community-linux-x86: [success](https://vulcan.livecode.com/builders/try-community-linux-x86/builds/414)
- try-community-mac: [success](https://vulcan.livecode.com/builders/try-community-mac/builds/419)
- try-community-linux-x86_64: [success](https://vulcan.livecode.com/builders/try-community-linux-x86_64/builds/414)
- try-community-emscripten: [success](https://vulcan.livecode.com/builders/try-community-emscripten/builds/414)
- try-community-ios: [success](https://vulcan.livecode.com/builders/try-community-ios/builds/420)
- try-community-win-x86: [success](https://vulcan.livecode.com/builders/try-community-win-x86/builds/414)
- try-community-android-armv6: [success](https://vulcan.livecode.com/builders/try-community-android-armv6/builds/414)

montegoulding commented on engine/src/uidc.cpp:
```diff
@@ -1272,36 +1328,40 @@ Boolean MCUIDC::handlepending(real8& curtime, real8& eventtime, Boolean dispatch
+            
+            // A message has been removed from the queue, so don't increment the
+            // counter on this iteration.
+            i -= 1;
```
@livecodefraser @peter-b doesn't `break` cover this?
